dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/CDIContainerImpl.java

```diff
@@ -593,17 +593,23 @@ public String getCurrentApplicationContextID() {
         //Now do the exact same thing for extensions coming from the SPI
         Iterator<ServiceAndServiceReferencePair<CDIExtensionMetadata>> spiExtensions = cdiRuntime.getSPIExtensionServices();
         while (spiExtensions.hasNext()) {
-            ServiceAndServiceReferencePair<CDIExtensionMetadata> extension = spiExtensions.next();
-            ServiceReference<CDIExtensionMetadata> sr = extension.getServiceReference();
+            ServiceAndServiceReferencePair<CDIExtensionMetadata> extensionMetaData = spiExtensions.next();
+            ServiceReference<CDIExtensionMetadata> sr = extensionMetaData.getServiceReference();
             if (sr != null) {
                 Long serviceID = ServiceReferenceUtils.getId(sr);
                 ExtensionArchive extensionArchive = null;
                 synchronized (this) {
                     extensionArchive = runtimeExtensionMap.get(serviceID);
 
                     if (extensionArchive == null) {
-                        extensionArchive = newSPIExtensionArchive(sr, extension.getService(), applicationContext);
+                        extensionArchive = newSPIExtensionArchive(sr, extensionMetaData.getService(), applicationContext);
                         runtimeExtensionMap.put(serviceID, extensionArchive);
+                    } else {
+                        // We don't need to create the extension archive but we do need to register the contained extensions into the deploymnet.
+                        // Since ExtensionArchive doesn't have a method for getting the extension classes, we'll get the extensions from the service.
+
+                        Set<Extension> spiExtensionClasses = getExtensionInstancesFromService(extensionMetaData.getService());
+                        applicationContext.registerSPIExtensions(spiExtensionClasses);
```

Comment:
Can you try to make the two paths a little closer? e.g. surely you can get the `spiExtensionClasses` from the extensionArchive? So the code would more more like this;

```
extensionArchive = runtimeExtensionMap.get(serviceID);
if (extensionArchive == null) {
    //only create a new archive, don't register any classes with the deployment
    extensionArchive = newSPIExtensionArchive(sr, extension.getService(), applicationContext);
    runtimeExtensionMap.put(serviceID, extensionArchive);
}
Set<Extension> spiExtensionClasses = extensionArchive.getExtensions();
applicationContext.registerSPIExtensions(spiExtensionClasses);
```

Reply:
oh actually, don't call it `spiExtensionClasses` because they are instances